### PR TITLE
Get basic functionality working on the NRF51

### PR DESCRIPTION
**Pull request by phil-levis**
The nRF51 now has a working GPIO and timer system. There's a copy of Utkarsha's UART but it is not yet tested. To see a one-shot timer working, use the blink_periodic app.

This also changes naming conventions, to follow the evolution of the nRF SoCs. chips/nrf51822 is now chips/nrf51; platforms/nrf_pca10001 (the nRF51822 EK) is now platforms/nrf_pca10028 (the nRF51 DK). The assumption is that we'll keep one chips/nrf51 and handle errata as well as differences between iterations within files.

Also, currently the timer system uses nRF51 Timer1: it should use the RTC. Similarly, systick should be rewritten to use Timer1.

Remaining to do items:
  - make systick based off TIMER1
  - make timer system based off RTC
  - test/finish UART and console

### CODE REVIEW

**Comment by alevy** [comment]
Can you remove the ble_stack in apps? It's most of the files in the pull request, but isn't really part of it. It makes the pull request very difficult to review (it literally breaks GitHub's "File's Changed" tab)

**Comment by phil-levis** [comment]
Done

Phil

File: apps/blink/main.c [hunk 1]
```diff
@@ -2,11 +2,13 @@
...
 int main(void) {
-    gpio_enable_output(LED_0);
-
+    gpio_set(0);
     while(1) {
-      gpio_toggle(LED_0);
-      delay_ms(500);
+      gpio_toggle(0);
+//      printf("Before delay.\n");
+      delay_ms(1000);
```
Comment (alevy) [inline apps/blink/main.c]: The changes to this file are _OK_, but I don't think there's any reason for them. Please at least remove the commented out code.

File: apps/blink_periodic/main.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+#include <firestorm.h>
+#include <gpio.h>
+
+/* Delay for for the given microseconds (approximately).
+ *
+ * For a 16 MHz CPU, 1us == 16 instructions (assuming each instruction takes
+ * one cycle). */
+static void busy_delay_us(int duration) {
```
Comment (alevy) [inline apps/blink_periodic/main.c]: Unused

File: apps/blink_periodic/main.c [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+				"nop\n"
+			       );
+	}
+}
+
+/* Delay for for the given milliseconds (approximately).
+ *
+ * Note that this is not precise as there are 2 extra instructions on the in    ner
+ * loop. Therefore, there is 1us added every 8 iterations. */
+static void busy_delay_ms(int duration)
```
Comment (alevy) [inline apps/blink_periodic/main.c]: Unused

File: apps/blink_periodic/main.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	for (int i = 0; i < 4; i++) {
+		gpio_enable_output(i);
+		gpio_set(i);
+	}
+	bool c = false;
+	while (1) {
+		gpio_toggle(0);
+		timer_subscribe(delay_cb, &c);
+		timer_oneshot(500);
+		wait_for(&c);
```
Comment (alevy) [inline apps/blink_periodic/main.c]: This app is actually functionally identical to `blink` (the `delay` function from the "becoming-standard-library" is just a wrapper around `timer_oneshot` that does exactly this).

What the name of the app implies is that it will use `timer_start_repeating` and not run in a while loop (but just return from main to the implied `while(1) { wait() }`)

File: apps/c_gpio/main.c [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    gpio_set(LED_1);
+    // Application pin 3 is HW GPIO pin 0, see nrf_pc10001/lib.rs
+    gpio_enable_input(3, PullDown);
+    gpio_enable_interrupt(3, PullDown, RisingEdge);
+    gpio_interrupt_callback(interrupt_callback, NULL);
+
+    int i;
+    for (i = 0; i < 10; i++) {
+      gpio_toggle(LED_0);
+      busy_delay_ms(200);
```
Comment (alevy) [inline apps/c_gpio/main.c]: Since timers presumably work, it would be better to use vanilla `delay` now

File: apps/libs/firestorm.c [hunk 6]
```diff
@@ -84,6 +84,7 @@ static void delay_cb( __attribute__ ((unused)) int unused0,
                       __attribute__ ((unused)) int unused1,
                       __attribute__ ((unused)) int unused2,
                       void* ud) {
+  gpio_toggle(3);
```
Comment (alevy) [inline apps/libs/firestorm.c]: please remove

File: golden/nrf-golden.hex [hunk 7]
```diff
@@ -0,0 +1,656 @@
+:10000000883A0020650100006101000061010000E4
```
Comment (alevy) [inline golden/nrf-golden.hex]: please remove this file, I don't think it's something we want in `master`

File: src/arch/cortex-m0/syscalls.S [hunk 8]
```diff
@@ -13,12 +13,25 @@
 .thumb_func
```
Comment (alevy) [inline src/arch/cortex-m0/syscalls.S]: All of these changes are now incorrect. The kernel is responsible for saving `r4-r7` (when we preempt we need to save them anyway, and in that case we don't get the application's cooperation)

File: src/chips/nrf51/Makefile.mk [hunk 9]
```diff
@@ -1,7 +1,7 @@
 ARCH = cortex-m0
 
 RUSTC_FLAGS += -C opt-level=3 -Z no-landing-pads
-RUSTC_FLAGS += --target $(SRC_DIR)chips/$(CHIP)/makefile_target.json
+RUSTC_FLAGS += --target $(SRC_DIR)chips/$(CHIP)/target.json
```
Comment (alevy) [inline src/chips/nrf51/Makefile.mk]: neither `makefile_target` nor `target` will work for both `Makefile` based compilation from root and Cargo based compilation without fixing the relative locations of the two crates because of the where the linker script is expected to be. In the transition period where we're still doing both (maybe not for long?) I suggest we just keep both copies.

It works on the SAM4L port because we don't have to compile directly an ELF with Rust, since we are not incorporating any native-C code. But that's not the case with the NRF51 because of the code in `crt1.c`.

File: src/chips/nrf51/src/clock.rs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+pub static mut CLOCK : Clock = Clock {
+    client: TakeCell::empty(),
+};
+
+
+impl Clock {
+    #[allow(non_snake_case)]
+    fn CLOCK(&self) -> &'static Registers {
```
Comment (alevy) [inline src/chips/nrf51/src/clock.rs]: perhaps a better name would be `regs()` or `registers()`? Alternatively, you could also adopt @lizardo's style of having a global `CLOCK()` function rather than a method. Finally, @KBaichoo and I have been experimenting with a special `StaticRef` type (see an example [here](https://github.com/alevy/tock-hotel/blob/driver/usb/src/chips/hotel/src/usb/mod.rs#L11)) but I don't suggesting using it yet as it's still pretty experimental.

File: src/platform/nrf_pca10028/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+pub struct Platform {
+    //chip: nrf51::chip::Nrf51,
+    gpio: &'static drivers::gpio::GPIO<'static, nrf51::gpio::GPIOPin>,
+    timer: &'static TimerDriver<'static, VirtualMuxAlarm<'static, TimerAlarm>>,
+    console: &'static drivers::console::Console<'static, nrf51::uart::UART>,
+}
+
+/* pub struct DummyMPU;
```
Comment (alevy) [inline src/platform/nrf_pca10028/src/main.rs]: remove

File: src/platform/nrf_pca10028/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+
+impl DummyMPU {
+    pub fn set_mpu(&mut self, _: u32, _: u32, _: u32, _: bool, _: u32) {
+    }
+}*/
+
+impl main::Platform for Platform {
+  /*  pub fn mpu(&mut self) -> DummyMPU {
```
Comment (alevy) [inline src/platform/nrf_pca10028/src/main.rs]: remove

File: src/platform/nrf_pca10028/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+    nrf51::clock::CLOCK.high_stop();
+
+    nrf51::clock::CLOCK.low_set_source(nrf51::clock::LowClockSource::RC);
+    nrf51::clock::CLOCK.low_start();
+    nrf51::clock::CLOCK.high_start();
+    while !nrf51::clock::CLOCK.low_started() {}
+    while !nrf51::clock::CLOCK.high_started() {}
+
+    static_init!(platform: Platform = Platform {
+        //chip: nrf51::chip::NRF51::new(),
```
Comment (alevy) [inline src/platform/nrf_pca10028/src/main.rs]: remove

File: src/platform/nrf_pca10028/src/systick.rs [hunk 14]
```diff
@@ -0,0 +1,134 @@
+//! Stub implementation of a systick timer since the NRF doesn't have the
```
Comment (alevy) [inline src/platform/nrf_pca10028/src/systick.rs]: This belongs in the `nrf51` crate instead. That way, the implementation for the `main::Chip` trait's `SysTick` associated type. At the moment this isn't actually doing anything :(

File: src/platform/storm/Makefile.mk [hunk 15]
```diff
@@ -14,19 +14,19 @@ PLATFORM_DEPS += $(BUILD_PLATFORM_DIR)/libmain.rlib
 all: $(BUILD_PLATFORM_DIR)/kernel.elf
```
Comment (alevy) [inline src/platform/storm/Makefile.mk]: please revert the changes to this file

**Comment by alevy** [comment]
Awesome job Phil!

At a high level this looks fine to merge with minor modifications:

1. The continuous build is broken. I believe it's just because the `.travis.yml` script is trying to build in the old `nrf_pca10001`.

2. A few extraneous files and other nits commented inline

3. If at all possible, it would be great if the commit log could be cleaned up from 89 commits to something more readable, but if it's too much trouble it's too much trouble (it might be with all the intermediate merging we had to do).

After addressing 1 and 2 (at least the more unquestionable parts of 2) we can merge happily and promptly.

File: apps/c_gpio/main.c (hunk 5 above)
Comment (phil-levis) [inline apps/c_gpio/main.c]: The goal of this app was to test GPIO and only GPIO, yet remain visible to a human. I'd rather not add an additional dependency (i.e., if timers break you can't test this app).

File: src/chips/nrf51/Makefile.mk (hunk 9 above)
Comment (phil-levis) [inline src/chips/nrf51/Makefile.mk]: So I should just make a copy of target.json and call it makefile_target.json?

File: src/platform/nrf_pca10028/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+    let total_size = volatile_load(addr as *const usize);
+    if total_size != 0 {
+        volatile_store(&mut PROCS[0], Some(main::process::Process::create(addr,
+                                                                          total_size, &    mut MEMORY)));
+    }
+    &mut PROCS
+}
+
+pub struct Platform {
+    //chip: nrf51::chip::Nrf51,
```
Comment (phil-levis) [inline src/platform/nrf_pca10028/src/main.rs]: Why? The SAM4L has this code. Is this because the NRF can currently only support a single process? What if that changes? It seemed like keeping the code as parallel as possible was desirable.

Comment (alevy) [inline src/platform/nrf_pca10028/src/main.rs]: Oh sorry, I was only referring to the commented out `chip` field

File: apps/c_gpio/main.c (hunk 5 above)
Comment (alevy) [inline apps/c_gpio/main.c]: Oh I see. That makes sense. Maybe we should have a better longer term solution for doing these kinds of tests vs having example apps, but that's not for a different PR and discussion altogether. This is fine as is

File: src/chips/nrf51/src/clock.rs (hunk 10 above)
Comment (phil-levis) [inline src/chips/nrf51/src/clock.rs]: It was a struct function because of the name conflict (CLOCK/CLOCK). To make it consistent with other implementations I made it global and renamed it CLK().

WRT StatifRef, avoid the Island Syndrome!

Comment (alevy) [inline src/chips/nrf51/src/clock.rs]: WRT StaticRef:

We haven't converged on a way of properly lifting pointers to register structures in Rust references. The purpose of `StaticRef` is to figure out how to require significantly less code to be trusted, where so far the methods we've been using haven't covered 100% of cases:

  * Throughout the SAM4L codebase we use `*const Registers` fields in the device struct and cast explicitly in each method. This works across devices with a single register base (e.g. the AST) as well as those with multiple (e.g. USART0, USART1, etc). However, it requires an unsafe cast in each method.

  * In the NRF51 @lizardo and you seem to be using the `DEVICE_NAME()` style, which isolates the `unsafe` code, but won't work as elegantly for device types with multiple instances (not sure if there are any on the nrf51).

  * StaticRef is the (almost) best of both worlds (it technically requires an extra word of memory for singleton devices relative to the second option above), but I haven't tested a full device with it yet.

In these cases, I think it's fine to try an isolated example. If it doesn't work out well, it's trivial to port it to the other mechanisms, and I will. If it does, and we decide to adopt it, I'll go through and change everything at once (also trivial, just touches a lot of files).

Comment (phil-levis) [inline src/chips/nrf51/src/clock.rs]: This is exactly the kind of documentation that such a type, if it emerges, needs. The island syndrome can be greatly reduced with good explanations.

Phil
